### ISSUE TITLE
吴恩达机器学习—03^单变量线性回归

### ISSUE THREAD

**Issue by noonbiteun** [noonbiteun/Blog#9]
**回归**的目的是为了预测，比如预测明天的天气温度，预测股票的走势…

回归之所以能预测是因为他通过历史数据，摸透了“套路”，然后通过这个套路来预测未来的结果。

![2019-12-27 16-50-02屏幕截图](https://user-images.githubusercontent.com/23396985/71510207-07a4cd80-28c9-11ea-86f7-890c8dd932ef.png)

**线性回归**本来是是统计学里的概念，现在经常被用在机器学习中。

如果 2 个或者多个变量之间存在“线性关系”，那么我们就可以通过历史数据，摸清变量之间的“套路”，建立一个有效的模型，来预测未来的变量结果。

![2019-12-27 16-51-11屏幕截图](https://user-images.githubusercontent.com/23396985/71510234-1ee3bb00-28c9-11ea-8a8d-cc22cf247bc5.png)

## 1 模型表示（Model Representation）

让我们以预测住房价格的例子开始：首先要使用一个数据集，数据集包含不同房屋尺寸所售出的价格，根据数据集画出我的图表。比方说，如果你朋友的房子是 1250 平方尺大小，你要告诉他们这房子能卖多少钱。那么，你可以做的一件事就是构建一个模型，也许是条直线，从这个数据模型上来看，也许你可以告诉你的朋友，他能以大约220000(美元)左右的价格卖掉这个房子。这就是监督学习算法的一个例子。

![1565629-20190113164351418-134224119](https://user-images.githubusercontent.com/23396985/71509421-2bb2df80-28c6-11ea-9e97-9b705d789177.png)

需要做的就是利用数据集中的数据，找出可以描述该数据集中输入和输出的关系的数学模型。在程序中我们要首先设定这么一个函数h（假设函数）

![2019-12-27 16-27-08屏幕截图](https://user-images.githubusercontent.com/23396985/71509288-c7901b80-28c5-11ea-8961-8dc71194805a.png)

对于只有一个输入的问题，我们可以将h表达为：![gif latex](https://user-images.githubusercontent.com/23396985/71509435-366d7480-28c6-11ea-8987-a836fe72408f.gif)

**只含有一个特征/输入变量的问题叫做单变量线性回归问题**

## 2 代价函数（Cost Function）

由上节可知，问题的核心在于找到一种可以描述数据集中输入和输出的关系的数学模型，即求解假设函数ℎ𝜃(𝑥) = 𝜃0 + 𝜃1𝑥。求解假设函数要做的就是为该函数选择合适的参数（parameters）𝜃0 和 𝜃1。我们选择的参数决定了我们得到的直线相对于我们的训练集的准确程度，模型所预测的值与训练集中实际值之间的差距（下图中蓝线所指）就是建模误差（modeling error）

![1565629-20190113170154386-702397076](https://user-images.githubusercontent.com/23396985/71509665-0e324580-28c7-11ea-8286-c61eccbf05cc.png)

**我们的目标便是选择出可以使得建模误差的平方和能够最小的模型参数**；
**即使得代价函数![1565629-20190113170417250-1716905159](https://user-images.githubusercontent.com/23396985/71509713-40dc3e00-28c7-11ea-879f-dcbd752233b2.png)最小**
找到了最小的代价函数，就找到了最能准确表达输入和输出之间关系的数学模型。

对于代价函数J，我们可以绘制出选择不同的参数𝜃0 和 𝜃1时，J的等高线图。从图中可以看出，在三维空间中存在一个使得𝐽(𝜃0, 𝜃1)最小的点。

![1565629-20190113170909501-773672878](https://user-images.githubusercontent.com/23396985/71509896-f3140580-28c7-11ea-958c-cace60a13dae.png) ![1565629-20190113171306039-1512451287](https://user-images.githubusercontent.com/23396985/71509908-fa3b1380-28c7-11ea-84ba-55eacf5430e1.png)

但是，绘出图形然后人工的方法来读出这些点的数值，找到最小的J，显然是个很麻烦的办法。因此需要找到一种算法，可以自动找到能使代价函数J最小化的参数𝜃0和𝜃1的值。**梯度下降算法**就是一种较好的选择。

## 3 梯度下降（Gradient Descent）

梯度下降的思想：

* start with some parameters：𝜃0, 𝜃1;
* keep changing parameters (𝜃0, 𝜃1) to reduce J(𝜃0, 𝜃1);
* get a local minimum（局部最小值）

 批量梯度下降（batch gradient descent）算法的公式为：

![1565629-20190113173424318-9799989](https://user-images.githubusercontent.com/23396985/71512217-23f83880-28d0-11ea-9258-b27e3112710a.png)

其中𝑎是学习率（learning rate），它决定了我们沿着能让代价函数下降程度最大的方向向下迈出的步子有多大

![ef4227864e3cabb9a3938386f857e938](https://user-images.githubusercontent.com/23396985/71512649-222f7480-28d2-11ea-8d49-7f5a68e7312e.png)

**值得注意的是：parameters：𝜃0, 𝜃1, . . . , 𝜃𝑛要同时更新**

梯度下降的直观理解：

![image](https://user-images.githubusercontent.com/23396985/71656411-087caa00-2d76-11ea-8e68-018db1751ad9.png)

描述：对𝜃赋值，使得𝐽(𝜃)按梯度下降最快方向进行，一直迭代下去，最终得到局部最小值。其中𝑎是学习率（learning rate），它决定了我们沿着能让代价函数下降程度最大的方向向下迈出的步子有多大。

通过上图对梯度下降的过程进行理解：
1. 如果函数处于下降趋势，则代价函数J的偏导数为负数，减去一个负数等于加上一个正数，那么𝜃的值就会变大，随着𝜃的变大，代价函数J逐渐减小，达到了降低J的目的；
2. 如果函数处于上升趋势，则代价函数J的偏导数为正数，减去一个正数，那么𝜃的值就会变小，随着𝜃的变小，代价函数J逐渐减小，同样也达到了降低J的目的。

 再来看看学习率𝑎：
1. 若𝑎太小，那么数据的挪动非常细致，同样也非常慢，需要很多步才能到达最低点；
2. 如果𝑎太大，那么梯度下降法可能会越过最低点，甚至可能无法收敛，甚至发散。

将梯度下降应用在线性回归中，具体的程序实现为：

![image](https://user-images.githubusercontent.com/23396985/71656515-7628d600-2d76-11ea-90ec-31f2d997d3e6.png)

```python
#单变量线性回归中的梯度下降
for iter = 1:num_iters
　　theta = theta-alpha*(1/m)*X'*((theta'*X')'-y);   #求解参数𝜃
　　J_history(iter) = computeCost(X, y, theta);     #计算代价函数J
end
```
